### DISCUSSION TITLE
Rewrite NativeScript Properties for NS 8(?)

### DISCUSSION THREAD

**Discussion by NathanaelA** [NativeScript/rfcs#29] [1. Ideas & Discussions]
For anyone who has looked at the Nativescript properties file here:  https://github.com/NativeScript/NativeScript/blob/master/packages/core/ui/core/properties/index.ts 

These are a colossal pain to even work with.   They are a mix of closures captures, direct captures and some class/object properties, making following the code difficult in some cases and also make it impossible to update some values on the classes at runtime.   

In addition each property also waste extra memory as frequently they not only assign a `this.x = options.x` but a `const x = options.x` and some of them even capture and use the `options` object, meaning the value(s) passed into the constructor is in memory three times for several of the property types.

In the past capture'd variables were faster than object properties.   However, doing some tests recently in Chrome 88 it shows that 
https://jsben.ch/rD6WO
![image](https://user-images.githubusercontent.com/850871/108928669-29eac600-7608-11eb-99fe-fb15faa2469b.png)

Please note that all tests should be done in a Chromium type browser as v8 is the same engine that we use in the standard Android & iOS runtimes.

JSPerf shows that Capture is 7% slower than Properties, and Direct is 10% slower.

Block one is Class A extends Class B extends Class C, with Class A function accessing Class C properties (basically the worst case type scenario for accessing a property in a class).
Block two is a direct capture (not even any extends, so the fastest capture that can ever occur)
Block three is a indirect capture (again no extends, so again the fastest indirect capture that can occur)

The current "Properties" code base uses all three of these method, but almost all properties are used via direct or indirect captures, so based on the above benchmarks (which I've ran multiple times, with multiple benchmark sites) all come out pointing that v8 (which iOS and Android use) are much faster using class properties.

---

Proposal, I would like to volunteer to rewrite this chunk of code, to be more maintainable, modern and hopefully faster and less memory usage.   It would not change any other part of the system as the properties would expose the same interface as the existing properties.

---

Any questions, suggestions or issues you see with this?

**Comment by NathanaelA** [comment]
# Test code:
## Initialization
```
// This is a normal class, where we assign properties, and access them using "this."
class Prop {
  constructor(options) {
     this.valueX = options.valueX;
     this.valueY = options.valueY;
  }
  doWork() {
    return this.valueX * this.valueY;    
  }
}

// This is a class that extends Prop, when you call 
// new Second().doWork() v8 has to go up the property chain to find doWork
// Benchmark shows this is slightly faster than Third
class Second extends Prop {
    constructor(options) {
       super(options);
    }
}

// This is a class that extends Second which then extends Prop, 
// meaning it has to go up the property chain to access "this.valueX"  and "this.valueY" each run of the local doWork...
class Third extends Second {
    constructor(options) {
      super(options);
    }
    doWork() {
      return this.valueX * this.valueY;
    }
}

// This creates two capture variables "valueX" and "valueY" to use them in the doWork function
class Capture {
  constructor(options) {
     const valueX = options.valueX;
     const valueY = options.valueY;
     this.doWork = function() {
        return valueX * valueY;
     };
  }
}


// This uses a capture of the passed in "options" object, so it has to do a lookup 
// on the options object to get "valueX" & "valueY" in the doWork function
// This is the SLOWEST method because of the lookup on the captured options
class Direct {
   constructor(options) {
      this.doWork = function() {
         return options.valueX * options.valueY;
      };
   }
}
```

## Test Functions: 
**Block 1**
``` 
// This tests Class Third extends Class Second extends class Prop, so three levels of prototypes for the do work function
let total1=0;
for (let i=1,j=1;i<100;i++,j+=2) {
  const options = {valueX: i, valueY: j};
  let q = new Third(options);   // Optionally you can replace "Third" with "Prop" to see speed of direct this access vs prototype this access
  total1 += q.doWork();
}

console.log(total1);
```

**Block 2**
```
// This tests the Capture class
let total2=0;
for (let i=1,j=1;i<100;i++,j+=2) {
  const options = {valueX: i, valueY: j};
  let q = new Capture(options);
  total2 += q.doWork();
}

console.log(total2);
```

**Block 3**
```
// This tests the Direct capture class
let total3=0;
for (let i=1,j=1;i<100;i++,j+=2) {
  const options = {valueX: i, valueY: j};
  let q = new Direct(options);
  total3 += q.doWork();
}

console.log(total3);
```

**Comment by farfromrefug** [comment]
@NathanaelA i totally agree we need to rewrite this!
A few questions:
* it means it would not work with iOS JSC? I am asking as i am not seeing V8 iOS as ready for production and it does not seem to be happening tomorrow. So i want to make sure it would stay compatible with non v8
* i have not completely understood what part you want to rewrite.  Could you detail what you want to "rewrite" (like the Style object, on `Property` class?)

**Reply by NathanaelA** [reply]
1. It would remain compatible with JSC.  ;-)     In the distant past captures were much faster, however in the present the optimization time for almost all JS engines has been on direct member access of a class, so captures are no longer as optimized of a path.   

2. Basically anything that extends from the base property class, (used by a lot of components as these are what trigger notifications when values on a class change) would turn into real class's rather than a mix of class and captures.    So when you do `Button.style.left = 10`.   the `.left` is a property.   When you do  `Button.text = "Blah"` the `.text` is a property.    

So basically the class you extend from for pretty much every component property is what I'm discussing should be re-factored to become full classes and eliminate the capturing to speed up all access to all properties.
